clippy_lints/src/types.rs

```diff
@@ -215,11 +215,41 @@ declare_clippy_lint! {
     "redundant allocation"
 }
 
+declare_clippy_lint! {
+    /// **What it does:** Checks for Rc<T> and Arc<T> when T is a mutable buffer type such as String or Vec
+    ///
+    /// **Why is this bad?** Expressions such as Rc<String> have no advantage over Rc<str>, since
+    /// it is larger and involves an extra level of indirection, and doesn't implement Borrow<str>.
+    ///
+    /// While mutating a buffer type would still be possible with Rc::get_mut(), it only
+    /// works if there are no additional references yet, which defeats the purpose of
+    /// enclosing it in a shared ownership type. Instead, additionally wrapping the inner
+    /// type with an interior mutable container (such as RefCell or Mutex) would normally
+    /// be used.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    /// ```rust,ignore
+    /// # use std::rc::Rc;
+    /// fn foo(interned: Rc<String>) { ... }
+    /// ```
+    ///
+    /// Better:
+    ///
+    /// ```rust,ignore
+    /// fn foo(interned: Rc<str>) { ... }
+    /// ```
+    pub RC_BUFFER,
+    nursery,
```

Comment:
I think it might be okay to put this in `style` instead of `nursery` unless you're particularly worried that this lint is unreliable.

Reply:
The issue suggested putting it into nursery, but since this turns out to be a pretty simple lint, I think that's okay too.

Most of the other type lints in this module are either perf (Box<Vec<T>>, Rc<&usize>) or complexity (Vec<Box<T>>).  I'm leaning towards perf.

Reply:
> leaning towards perf

:+1: make it so

Reply:
Moving it out of nursery caused dogfood to fail due to the Constant enum in clippy_lints/src/consts.rs which is converted from [LitKind](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_ast/ast/enum.LitKind.html), and I'm not sure how to best proceed.

The easiest answers would be to either allow this lint through for that definition only, or to "fix" the enum there only and forcing a string copy.  I don't particularly like either of these answers, especially not the second one.

The better answer seems like it'd be to fix the type as its defined in rust first.  I tried making the [change locally](https://github.com/rschoon/rust/commit/83d3cc408168c3200c0b4cd632b1b3c6f8244156), and and it seems to work, but I don't know how these inter-dependencies between rust and clippy are actually supposed to be handled (would that just mean for some period of time, clippy would be broken by the change in rustc's internals? and this PR would just be blocked until then?)

Reply:
> The easiest answers would be to either allow this lint through for that definition only, or to "fix" the enum there only and forcing a string copy. I don't particularly like either of these answers, especially not the second one.

Assuming I'm understanding this correctly I'd lean towards just updating the definition of `ConstantBinary` and using the `From<Vec<T>>` impl on `Rc` to convert it inside of `lit_to_constant`. Then if you wanted you could do a subsequent PR to `rustc` where you fix the upstream type. I think you'd also be able to fix the clippy definition at the same time thanks to the new `git subtree` setup.

Reply:
Well, my concern is basically (afaik) we can't get from `Rc<Vec<u8>>` to `Rc<[u8]>` without copying the backing data, which we didn't have to do before.  However, I'm feeling a bit assuaged on this because we're actually already copying data on the `LitKind::Str` case (and that probably comes up a lot more).

I don't think `From<Vec<T>>` helps here, since we would need to get the `Vec<u8>` out of the Rc, and we're not the exclusive owner, but it's not the only way to get our copy.

Reply:
Actually, I guess I could have gotten a clone of the Vec out of the Rc.